### ISSUE TITLE
Possible problem in preparing alpaca data

### ISSUE THREAD

**Issue by marcecaro** [Lightning-AI/lit-llama#279]
according to the documentation of prepare_sample:

```
    This function processes this data to produce a prompt text and a label for
    supervised training. The prompt text is formed as a single message including both
    the instruction and the input. The label/target is the same message but with the
    response attached.
```

but the code shows that "input_ids" (which is used latter as X), and "labels", both contains promp and response, if not using making, then they are exactly the same:

```python
    # The labels are the full prompt with response, but with the prompt masked out
    labels = encoded_full_prompt_and_response.clone()
    if mask_inputs:
        labels[:len(encoded_full_prompt)] = IGNORE_INDEX

    return {**example, "input_ids": encoded_full_prompt_and_response, "input_ids_no_response": encoded_full_prompt, "labels": labels}
```

Latter during training, we are taking 'input_ids' and 'labels', which are the same:

```python
def get_batch(fabric: L.Fabric, data: list):
    ix = torch.randint(len(data), (micro_batch_size,))

    input_ids = [data[i]["input_ids"].type(torch.int64) for i in ix]
    labels = [data[i]["labels"].type(torch.int64) for i in ix]
```

Maybe I am missing something,

Cheers

**Comment by miso-choi** [comment]
I was also wondering about this. Looking forward to an explanation. 
Thank you for sharing this project :)

**Comment by Freddavide** [comment]
I encountered the same issue!

**Comment by lantiga** [comment]
Odd as it is, this is the way supervised fine-tuning works in this case.

Remember that the model is only capable of generating one token at a time, so it would be very impractical to generate a full response token by token and compare it with the target (and you could't simply use a single backward to optimize this process).

What happens is that you take inputs *and* outputs, give them in input, and then look at the logits generated for the output portion of the input (if masking is on, otherwise the full prompt directly) and take a gradient descent step to change the parameters so that those logits will be higher the next time around.

**Comment by awaelchli** [comment]
Adding to @lantiga's response, here is where the labels get used: 
https://github.com/Lightning-AI/lit-llama/blob/ee32d4fa7b663a0f34d894190b258a5138d7254b/finetune/adapter.py#L197-L199

The sequence gets shifted by 1 so that the loss will be between the predicted token for input t and the label token t+1. Whether this shifting should be done in the data preprocessing code or online is a stylistic choice. Masking the part before the response is optional, but could be done. If masking is not desired, you can simplify the code above by just returning input_ids = labels. Hope this helps.

**Comment by awaelchli** [comment]
Let us know if these clarifications help. I updated the comment in the code: https://github.com/Lightning-AI/lit-llama/pull/282

**Comment by marcecaro** [comment]
It helps! Thank you very much!!
